### PR TITLE
[XLA/GPU] Improve h-loop-fusion by changing its traversal order.

### PR DESCRIPTION
**Pull request by trentlo**
Switch to traverse from use to def. Bitcasts are placed after h-fusions to
resolve shape mismatch but bitcasts could prevent future h-fusion from
happening. So, a bottom-up, use-to-def order should be more favorable. It
also helps to save compiler iterations to reach the fixed point. See
HorizontalLoopFusionTest::TraversalOrder for an example.

We see ~7% gain from the LAMB training optimizer in a real model.

### CODE REVIEW

**Comment by trentlo** [comment]
@cheshire could you help to take a look of this? Thanks!

**Review by cheshire — APPROVED** [review]

File: tensorflow/compiler/xla/service/gpu/horizontal_loop_fusion.cc [hunk 1]
```diff
@@ -479,11 +479,15 @@ StatusOr<bool> HorizontalLoopFusionImpl::Run() {
...
+  // shape mismatch but bitcasts could prevent future h-fusion from happening.
+  // So, a bottom-up, use-to-def order should be more favorable. It also helps
+  // to save compiler iterations to reach the fixed point.
+  std::vector<HloInstruction*> use_to_def_order =
       computation_->MakeInstructionPostOrder();
-  for (size_t i = 0; i < def_to_use_order.size(); ++i) {
-    HloInstruction* consumer = def_to_use_order[i];
+  absl::c_reverse(use_to_def_order);
+  for (size_t i = 0; i < use_to_def_order.size(); ++i) {
+    auto consumer = use_to_def_order[i];
```
Comment (cheshire) [inline tensorflow/compiler/xla/service/gpu/horizontal_loop_fusion.cc:490]: Would prefer the actual type.

File: tensorflow/compiler/xla/service/gpu/horizontal_loop_fusion_test.cc [hunk 2]
```diff
@@ -505,6 +505,73 @@ TEST_F(HorizontalLoopFusionTest, IterativeHorizontalFusion) {
   EXPECT_EQ(total_fusion_instrs, 2);
 }
 
+TEST_F(HorizontalLoopFusionTest, TraversalOrder) {
+  auto module = ParseAndReturnVerifiedModule(R"(
```
Comment (cheshire) [inline tensorflow/compiler/xla/service/gpu/horizontal_loop_fusion_test.cc:509]: Also would prefer the actual type.

File: tensorflow/compiler/xla/service/gpu/horizontal_loop_fusion.cc (hunk 1 above)
Comment (trentlo) [reply tensorflow/compiler/xla/service/gpu/horizontal_loop_fusion.cc]: Make sense. Done.

File: tensorflow/compiler/xla/service/gpu/horizontal_loop_fusion_test.cc (hunk 2 above)
Comment (trentlo) [reply tensorflow/compiler/xla/service/gpu/horizontal_loop_fusion_test.cc:521]: Expanding this one is a bit lengthy. (i..e, `StatusOr<std::unique_ptr<VerifiedHloModule>>`)

Skip it but I expand all other reasonable candidates in the file though.

**Review by trentlo — COMMENTED** [review]
Please take a look again. Thanks!

**Review by cheshire — APPROVED** [review]

**Review by gbaned — APPROVED** [review]

**Pull request merged**